pkg/policy/groups/helpers.go

```diff
@@ -0,0 +1,152 @@
+		parentCNP:  string(cnp.ObjectMeta.UID),
+		cnpKindKey: cnpKindName,
+	}
+	derivativeCNP.Spec = &api.Rule{}
+	derivativeCNP.Specs = api.Rules{}
+
+	rules, err := cnp.Parse()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse policies: %s", err)
+	}
+
+	for _, rule := range rules {
+		if !rule.RequiresDerivative() {
+			continue
+		}
+		newRule, err := rule.CreateDerivative()
+		if err != nil {
+			return derivativeCNP, err
+		}
+		derivativeCNP.Specs = append(derivativeCNP.Specs, newRule)
+	}
+
+	return derivativeCNP, nil
+}
+
+// getK8sClient return the kubernetes apiserver connection
+func getK8sClient() (*clientset.Clientset, error) {
+	k8sMutex.Lock()
+	defer k8sMutex.Unlock()
+	if globalK8sClient != nil {
+		return globalK8sClient, nil
+	}
+
+	restConfig, err := k8s.CreateConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create rest configuration: %s", err)
+	}
+	k8sClient, err := clientset.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create Kubernetes configuration: %s", err)
+
+	}
+	globalK8sClient = k8sClient
+	return globalK8sClient, nil
+}
+
+func updateCNPStatus(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		// @TODO change the error here
+		return nil, err
+	}
+	return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).UpdateStatus(cnp)
+}
+
+func updateOrCreateCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		return nil, err
+	}
+	k8sCNP, err := k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).
+		Get(cnp.ObjectMeta.Name, v1.GetOptions{})
+	if err == nil {
+		cnp.ObjectMeta.ResourceVersion = k8sCNP.ObjectMeta.ResourceVersion
+		return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).Update(cnp)
+	}
+	return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).Create(cnp)
+}
+
+func updateDerivativeStatus(cnp *cilium_v2.CiliumNetworkPolicy, derivativeName string, err error) error {
+	status := cilium_v2.CiliumDerivativePolicyStatus{
+		LastUpdated: cilium_v2.NewTimestamp(),
+		Enforcing:   false,
+	}
+
+	if err != nil {
+		status.OK = false
+		status.Error = fmt.Sprintf("%v", err)
+	} else {
+		status.OK = true
+		status.Error = ""
+	}
+
+	k8sClient, clientErr := getK8sClient()
+	if clientErr != nil {
+		return fmt.Errorf("Cannot get Kubernetes apiserver client: %s", clientErr)
+	}
+	// This CNP can be modified by cilium agent or operator. To be able to push
+	// the status correctly fetch the last version to avoid updates issues.
+	k8sCNPStatus, clientErr := k8sClient.CiliumV2().
+		CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).
+		Get(cnp.ObjectMeta.Name, v1.GetOptions{})
+	if clientErr != nil {
+		return fmt.Errorf("Cannot get Kubernetes policy: %s", clientErr)
+	}
+	if k8sCNPStatus.ObjectMeta.UID != cnp.ObjectMeta.UID {
```

Comment:
Is it possible for the UID to be different? or is that the case where a new version of the CNP is uploaded. I'm not sure I understand the error case. We delete the CNP we fetched, which has a different UID from the local one. Isn't it more correct to delete the old one? (I might still need to go find the groupsCNPCache and how it works).
Relatedly, `k8sCNPStatus` was really confusing to me, since it was an actual CNP and not a status (which one can get, I think, via a getStatus call or something). upstreamCNP might be a better name but I'm not sure.

Reply:
Hi,

I didn't get this corner case, but I added that check just in case. The delete will happens on DeleteCNPIfNeeded, so no really needed, this is only a safe check. 

About delete it, I think that is worth to keep it, it'll make this easy to debug if a problem happens.